applications/crossbar/src/modules/cb_token_restrictions.erl

```diff
@@ -106,147 +125,148 @@ delete(Context) ->
-            lager:debug("match rule: ~p",[MatchRule]),
-            wh_json:get_value(MatchRule, R)
+-spec allowed_accounts(api_binary(), api_binary()) -> ne_binaries().
+-ifdef(TEST).
+allowed_accounts(?AUTH_ACCOUNT_ID, ?ACCOUNT_ID = AccountId) ->
+    [?CATCH_ALL, AccountId, <<"{DESCENDANT_ACCOUNT_ID}">>].
+-else.
+allowed_accounts('undefined', _AccountId) -> [?CATCH_ALL];
+allowed_accounts(_AuthAccountId, 'undefined') -> [?CATCH_ALL];
+allowed_accounts(AuthAccountId, AccountId) ->
+    case wh_util:is_in_account_hierarchy(AuthAccountId, AccountId) of
+        'true' -> [?CATCH_ALL, AccountId, <<"{DESCENDANT_ACCOUNT_ID}">>];
+        'false' -> [?CATCH_ALL, AccountId]
     end.
+-endif.
 
-match_rules(_ReqParam, []) -> 'undefined';
-match_rules(ReqParam, [R|Rules]) ->
-    Rule = binary:split(R, <<"/">>, [global, trim]),
-    case kazoo_bindings:matches(Rule, ReqParam) of
-        'true' -> R;
-        'false' -> match_rules(ReqParam, Rules)
+-spec match_arguments(cb_context:context(), api_object()) ->
+                             http_methods().
+match_arguments(_Context, 'undefined') -> [];
+match_arguments(Context, RulesJObj) ->
+    [{_, ReqParams}|_] = cb_context:req_nouns(Context),
+    RuleKeys = wh_json:get_keys(RulesJObj),
+    match_argument_patterns(ReqParams, RulesJObj, RuleKeys).
+
+-spec match_argument_patterns(req_nouns(), wh_json:object(), ne_binaries()) ->
+                                     http_methods().
+match_argument_patterns(_ReqParams, _RulesJObj, []) -> [];
+match_argument_patterns(ReqParams, RulesJObj, RuleKeys) ->
+    case match_rules(ReqParams, RuleKeys) of
+        'undefined' -> [];
+        MatchedRuleKey ->
+            wh_json:get_value(MatchedRuleKey, RulesJObj, [])
     end.
 
--spec match_verb(cb_context:context(), list()) -> list() | 'undefined'.
-match_verb(Context, Verbs) ->
-    Verb = cb_context:req_verb(Context),
-    case lists:member(Verb, Verbs) orelse lists:member(<<"_">>, Verbs) of
-        'true' ->
-            lager:debug("match verb: ~p", [Verb]),
-            [Verb];
-        'false' ->
-            lager:info("no match verb for: ~p", [Verb]),
-            'undefined'
+-spec match_rules(ne_binaries(), ne_binaries()) -> api_binary().
+match_rules(_ReqParams, []) -> 'undefined';
+match_rules(ReqParams, [RuleKey|RuleKeys]) ->
+    case does_rule_match(binary:split(RuleKey, <<"/">>, ['global', 'trim'])
+                         ,ReqParams
+                        )
+    of
+        'false' -> match_rules(ReqParams, RuleKeys);
+        'true' -> RuleKey
     end.
+
+-spec does_rule_match(ne_binaries(), ne_binaries()) -> boolean().
+does_rule_match(Rule, ReqParams) ->
+    kazoo_bindings:matches(Rule, ReqParams).
+
+-spec match_verb(cb_context:context(), http_methods()) -> boolean().
+match_verb(_Context, [?CATCH_ALL]) -> 'true';
+match_verb(Context, RuleVerbs) ->
+    lists:member(cb_context:req_verb(Context), RuleVerbs)
+        orelse lists:member(?CATCH_ALL, RuleVerbs).
```

Comment:
In theory, one would only use `["_"]` so we can short circuit `match_verb` in the first clause and avoid calls to `lists:member/2`. Otherwise we need to check that someone didn't include the catch-all in a list with other verbs.